File: JS(Unused)/grid.js
export class AreaGrid {
  constructor(areaName, rows, cols) {
    this.areaName = areaName;
    this.rows = rows;
    this.cols = cols;
    this.grid = Array.from({ length: rows }, () => Array(cols).fill(null));
  }

  place(entity, row, col) {
    if (row < 0 || row >= this.rows || col < 0 || col >= this.cols) return false;
    if (this.grid[row][col] !== null) return false;
    this.grid[row][col] = entity;
    return true;
  }

  remove(row, col) {
    if (row < 0 || row >= this.rows || col < 0 || col >= this.cols) return null;
    const entity = this.grid[row][col];
    this.grid[row][col] = null;
    return entity;
  }

  get(row, col) {
    if (row < 0 || row >= this.rows || col < 0 || col >= this.cols) return null;
    return this.grid[row][col];
  }
}

export const areaGrids = {
  garden: new AreaGrid("garden", 6, 6),
  fishpond: new AreaGrid("fishpond", 5, 5),
  livestock: new AreaGrid("livestock", 5, 6),
};

File: Python(DSA)/grid.py
"""
FarmGrid — 2D list (static rows × cols)
place / remove / get → O(1)
"""


class FarmGrid:
    """2D list; grid[row][col] holds None or an entity id/object."""

    def __init__(self, rows: int, cols: int):
        self.rows = rows
        self.cols = cols
        self.grid = [[None for _ in range(cols)] for _ in range(rows)]

    def place(self, entity, row: int, col: int) -> bool:
        if not (0 <= row < self.rows and 0 <= col < self.cols):
            return False
        if self.grid[row][col] is not None:
            return False
        self.grid[row][col] = entity
        return True

    def remove(self, row: int, col: int):
        if not (0 <= row < self.rows and 0 <= col < self.cols):
            return None
        entity = self.grid[row][col]
        self.grid[row][col] = None
        return entity

    def get(self, row: int, col: int):
        if not (0 <= row < self.rows and 0 <= col < self.cols):
            return None
        return self.grid[row][col]


if __name__ == "__main__":
    g = FarmGrid(3, 3)
    print(g.place("coop", 1, 1))   # True
    print(g.place("pond", 1, 1))   # False (occupied)
    print(g.get(1, 1))             # coop
    print(g.remove(1, 1))          # coop
    print(g.get(1, 1))             # None